### PR TITLE
decimals for rates

### PR DESCRIPTION
**Pull request by GusevTimofey**

### CODE REVIEW

File: modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/gateways/Metadata.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import sttp.model.Uri.Segment
+import tofu.Throws
+import tofu.logging.{Logging, Logs}
+import tofu.syntax.logging._
+import tofu.syntax.monadic._
+import fi.spectrumlabs.rates.resolver.models.{Metadata => Meta}
+
+trait Metadata[F[_]] {
+
+  def getTokenInfo(token: AssetClass): F[Option[Meta]]
```
Comment (oskin1) [inline modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/gateways/Metadata.scala:19]: Method name diverges from the return type.

File: modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/gateways/Metadata.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    backend: SttpBackend[F, _],
+    logs: Logs[I, F]
+  ): I[Metadata[F]] =
+    logs.forService[Network[F]].map(implicit __ => new Impl[F](config))
+
+  final private class Impl[F[_]: Monad: Throws: Logging](config: NetworkConfig)(implicit backend: SttpBackend[F, _])
+    extends Metadata[F] {
+
+    def getTokenInfo(token: AssetClass): F[Option[Meta]] =
+      info"Going to get info for token $token from metadata service" >>
```
Comment (oskin1) [inline modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/gateways/Metadata.scala:35]: Looks like an aspect

File: modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import cats.{Functor, Monad, Parallel}
+import fi.spectrumlabs.core.models.domain.AssetClass
+import fi.spectrumlabs.rates.resolver.{AdaAssetClass, AdaDecimal, AdaMetadata}
+import fi.spectrumlabs.rates.resolver.gateways.Metadata
+import fi.spectrumlabs.rates.resolver.models.{Metadata => Meta}
+import tofu.logging.Logs
+import tofu.syntax.monadic._
+
+trait MetadataService[F[_]] {
+  def getTokensInfo(tokens: List[AssetClass]): F[List[Meta]]
```
Comment (oskin1) [inline modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala:14]: Looks like this method should be in one module with getTokenInfo. Models are the same, so Metadata and MetadataServer distinction doesn't make sense. 
You can make Metadata a pure client whose methods return models as-is without any conversions. Then do all the conversions in the MetadataService.

**Review by oskin1 — CHANGES_REQUESTED** [review]

File: modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala (hunk 3 above)
Comment (GusevTimofey) [reply modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala:14]: I wanna do it like this:
Client is `trait Meta[F[_]] { def getAssetMeta(asset: AssetClass): F[Response] }` and service is `trait MetaService[F[_]] { def getMetaForAssets(in: List[AssetClass]): F[List[AssetMeta]] }`

File: modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    logs: Logs[I, F]
+  ): I[MetadataService[F]] =
+    logs.forService[MetadataService[F]].map(implicit __ => new Impl[F])
+
+  final private class Impl[F[_]: Monad: Parallel](implicit meta: Metadata[F]) extends MetadataService[F] {
+
+    def getTokensMeta(tokens: List[AssetClass]): F[List[Meta]] =
+      tokens.distinct
+        .filterNot(_ =!= AdaAssetClass)
+        .parTraverse(asset => meta.getTokenMeta(asset).map(_.map(r => Meta(r.decimals, asset))))
```
Comment (oskin1) [inline modules/rates-resolver/src/main/scala/fi/spectrumlabs/rates/resolver/services/MetadataService.scala:31]: tofu.syntax.foption mapIn

**Review by oskin1 — APPROVED** [review]

**Pull request merged**